### DISCUSSION TITLE
Make it possible to simply import ib_async (or eventkit) without an event loop running

### DISCUSSION THREAD

**Discussion by fredmonroe** [ib-api-reloaded/ib_async#62] [General]
hey everyone - what's the status of eventkit which ib_async relies upon?
Is that part of the forking too?

I'd like to change eventkit (and thus ib_async) so that merely the act of importing it doesn't cause a failure if there is no event loop existing. (so i don't have to always have to do nest_async.apply())

currently this line:

https://github.com/erdewit/eventkit/blob/caec9382bf4e8d7ebf12e7772b7feaa39f930f74/eventkit/util.py#L24

main_event_loop = get_event_loop()

in eventkit.utils declares a global variable that fails if there is no running event loop on import

I have code that looks at ib_async types to automatically build database orm objects and so. If i'm in a notebook, or just generally don't have a running even loop, just as simple import ib_async will raise an exception

i think it would be trivial to fix this in eventit. Is there a preferred fork I could submit a PR to?

**Comment by mattsta** [comment]
I don't think we've seen that causing problems before. What error does it cause?

You are right it shouldn't be doing things globally though.

I can seem to import and test the event loop state even if nothing is running:

```python
>>> import asyncio
>>> f = asyncio.get_event_loop_policy().get_event_loop()
>>> f.is_running()
False
```

We don't have an active fork of `eventkit` to submit updates to, so we'd need a slightly larger effort to fork -> new pypi package -> update dependencies across a couple places.

**Reply by gnzsnz** [reply]
@mattsta I offer as volunteer to maintain eventkit. would you be ok to include it on ib-api-reloaded organization?

I will trigger PEP 541 for it, so we can keep the package name

**Reply by mattsta** [reply]
oh thanks! I'll fork it here this evening (and maybe give it a code formatter pass and get it using poetry for deps) then it should be good to continue cleaning up.

**Reply by gnzsnz** [reply]
> oh thanks! I'll fork it here this evening (and maybe give it a code formatter pass and get it using poetry for deps) then it should be good to continue cleaning up.

once the repository is in place i will trigger PEP 541, hope we can get the package name quickly, we meet all requirements for PEP 541

@mattsta  do you mind doing the same for [nest asyncio](https://github.com/erdewit/nest_asyncio)? ie fork the repository and grant me access to maintain it? again, i will trigger PEP 541,

**Reply by fredmonroe** [reply]
here is what i see when i do:
`import ib_async` inside ipython, ending in an exception `RuntimeError: There is no current event loop in thread 'MainThread'.`

I can avoid this by using nested_async, but i'd prefer not to have to do that just for an import, only if i actually call things that use the event loop.

So for example if i want to parse an xml flexreport, i don't want to have to create an event loop

```
Python 3.12.3 (main, Apr 27 2024, 19:00:21) [GCC 11.4.0]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.26.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import ib_async
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[1], line 1
----> 1 import ib_async

File ~/git/polysail/.venv/lib/python3.12/site-packages/ib_async/__init__.py:6
      3 import dataclasses
      4 import sys
----> 6 from eventkit import Event
      8 from . import util
      9 from .client import Client

File ~/git/polysail/.venv/lib/python3.12/site-packages/eventkit/__init__.py:3
      1 """Event-driven data pipelines."""
----> 3 from .event import Event
      4 from .ops.aggregate import (
      5     All, Any, Count, Deque, Ema, List, Max, Mean, Min, Pairwise, Product,
      6     Reduce, Sum)
      7 from .ops.array import (
      8     Array, ArrayAll, ArrayAny, ArrayMax, ArrayMean, ArrayMin, ArrayStd,
      9     ArraySum)

File ~/git/polysail/.venv/lib/python3.12/site-packages/eventkit/event.py:9
      4 import weakref
      5 from typing import (
      6     Any as AnyType, AsyncIterable, Awaitable, Iterable, List, Optional,
      7     Tuple, Union)
----> 9 from .util import NO_VALUE, get_event_loop, main_event_loop
     12 class Event:
     13     """
     14     Enable event passing between loosely coupled components.
     15     The event emits values to connected listeners and has
   (...)
     19         name: Name to use for this event.
     20     """

File ~/git/polysail/.venv/lib/python3.12/site-packages/eventkit/util.py:24
     20     """Get asyncio event loop, running or not."""
     21     return asyncio.get_event_loop_policy().get_event_loop()
---> 24 main_event_loop = get_event_loop()
     27 async def timerange(start=0, end=None, step: float = 1) \
     28         -> AsyncIterator[dt.datetime]:
     29     """
     30     Iterator that waits periodically until certain time points are
     31     reached while yielding those time points.
   (...)
     46             to space between values.
     47     """

File ~/git/polysail/.venv/lib/python3.12/site-packages/eventkit/util.py:21, in get_event_loop()
     19 def get_event_loop():
     20     """Get asyncio event loop, running or not."""
---> 21     return asyncio.get_event_loop_policy().get_event_loop()

File /usr/lib/python3.12/asyncio/events.py:702, in BaseDefaultEventLoopPolicy.get_event_loop(self)
    699     self.set_event_loop(self.new_event_loop())
    701 if self._local._loop is None:
--> 702     raise RuntimeError('There is no current event loop in thread %r.'
    703                        % threading.current_thread().name)
    705 return self._local._loop

RuntimeError: There is no current event loop in thread 'MainThread'.
```

**Reply by mattsta** [reply]
Thanks for the full log detail.

It looks like it's actually maybe a jupyter/ipython problem. I can run the event loop fetch with _no error_ in ipython 8.14.0, but in ipython 8.26.0 I get the `RuntimeError` (and there is no `RuntimeError` when running in regular `python` REPL itself).

So looks like it just started breaking over time 🤷‍♂️

We'll get that global removed and get dependencies updated soon.

**Reply by gnzsnz** [reply]
We discussed this on `climbermel` discord, there is a combination of python/ipython that causes this issue. 

```python
Jupyter QtConsole 5.5.2
Python 3.12.3 (main, Jul 30 2024, 13:46:48) [Clang 15.0.0 (clang-1500.3.9.4)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.26.0 -- An enhanced Interactive Python. Type '?' for help.

import ib_async

import asyncio

asyncio.get_event_loop_policy().get_event_loop().is_running()
Out[3]: True
```
I used to have it but now is working fine for me, i don't know what's the exact combination. I'm using 8.26.0 like you, but is working fine for me. doing `nest_asyncio.apply()` works

And upgrading python/ipython will most probably solved it. it did it for me.

**Comment by gnzsnz** [comment]
I have requested the transfer

https://github.com/pypi/support/issues/4563
https://github.com/pypi/support/issues/4562

**Comment by gnzsnz** [comment]
@fredmonroe  could you please check https://github.com/ib-api-reloaded/eventkit/issues/2

you can pull from [get_event_loop](https://github.com/ib-api-reloaded/eventkit/tree/get_event_loop) branch and test the issue. i can't reproduce it. from my side it's working fine

![image](https://github.com/user-attachments/assets/21ae5811-31d1-4e01-a648-3e5278db52bd)

**Comment by fredmonroe** [comment]
Ok, I will check.

**Comment by fredmonroe** [comment]
In case you didn't see my note on the PR, this fix works for me in my
environment - thanks for the change.

**Reply by gnzsnz** [reply]
thanks for your confirmation @fredmonroe 

I will add a test for `Event.emit_threadsafe` to confirm everything works properly.

For the moment there is not much to do. once we have a response from pypi i will push and upgraded package.

@mattsta , ib_async is not using `Event.emit_threadsafe` so in principle no impact at all. let me know if you would like to do a smoke test before pushing eventkit. From my side i don't see any need to rush this.
